testRequirement.coffee

```diff
@@ -62,35 +70,74 @@ matchesSubstring = (objectKey, substringTest, object) ->
   substringValues = substringTest['substring'] # array of test substrings
   substringValues = if util.isArray substringValues then substringValues else [ substringValues ] #put substringValues in array if it isn't already
 
-  objectValue = object[objectKey] # array of document field values
+  objectValue = object[objectKey] # document field values
   objectValues = if util.isArray objectValue then objectValue else [ objectValue ] #put objectValues in array if it isn't already
 
   for objectValue in objectValues
+    objectValueStr = JSON.stringify(objectValue) # look for substrings in stringified arrays and objects too
+    match = false
     for substring in substringValues
       if 'ignoreCase' not of substringTest or substringTest['ignoreCase'] # ignore case of string by default and when ignoreCase is true
         re = RegExp substring, "im" # case insensitive, multiline matching
       else
         re = RegExp substring, "m" # multiline matching
-      if objectValue
-        objectValue = objectValue.match re
+      if objectValueStr
+        match = re.test objectValueStr
 
-      # if substring in object value
-      return true if objectValue
+      # if substring in object value str
+      return true if match
+  return false
+
+# Check if the test is an 'empty' test
+isRunEmptyTest = (testValues) ->
+  if (_.isPlainObject testValues) and 'empty' of testValues then true else false
+
+# Check if specific field value is empty/non-empty according to expected test boolean
+# a field value is empty if it is "", [], {}, null, 0
+matchesEmptyTest = (objectKey, emptyTest, object) ->
+  if objectKey not of object
+    return false
+
+  objectValue = object[objectKey]
+  shouldBeEmpty = emptyTest['empty']
+
+  if typeof shouldBeEmpty isnt 'boolean'
+    return false
+
+  if shouldBeEmpty #return true if objectValue is empty
+    return (objectValue is null) or (objectValue is 0) or (objectValue.length is 0) or (Object.keys(objectValue).length is 0)
+
+  else #return true if objectValue is not empty
+    return not ((objectValue is null) or (objectValue is 0) or (objectValue.length is 0) or (Object.keys(objectValue).length is 0))
```

Comment:
You can merge this and the above if statment by doing checking if isEmpty. Then return `if shouldBeEmpty return isEmpty else return not isEmpty`